src/lua/protobuf.lua

```diff
@@ -0,0 +1,316 @@
+            end
+        end
+        declarations[field_name] = true
+        return_protocol[field_name] = def
+    end
+    for field_type, val in pairs(declarations) do
+        if not val then
+            error(('Type %q is not declared'):format(field_type))
+        end
+    end
+    setmetatable(return_protocol, protocol_mt)
+    return return_protocol
+end
+
+local function check_integer(data_spec, value)
+    if type(value) == 'number' and math.ceil(value) ~= value then
+        error(('Input number value %f for %q is not integer'):format(
+            value, data_spec['name']))
+    elseif (type(value) == 'cdata' and
+            not c_integer[tostring(ffi.typeof(value))]) then
+        error(('Input cdata value %q for %q field is not integer'):format(
+            tostring(ffi.typeof(value)), data_spec['name']))
+    end
+end
+
+local function code(data_spec, value)
+    local data_type = data_spec['type']
+    if (protobuf_varint_types[data_type] and
+        input_varint_types[type(value)]) then
+        check_integer(data_spec, value)
+        return encoder.encode_integer(data_spec, value)
+    elseif (protobuf_len_types[data_type] and
+            input_len_types[type(value)]) then
+        return encoder.encode_len(data_spec, value)
+    elseif (data_type == 'float' and type(value) == 'number') then
+        return encoder.encode_I32(data_spec, value)
+    elseif (data_type == 'double' and type(value) == 'number') then
+        return encoder.encode_I64(data_spec, value)
+    elseif (protobuf_I32_types[data_type] and
+            input_I32_types[type(value)]) then
+        if (type(value) == 'number' and (value > max_int32-1 or
+                value < -max_int32)) then
+            error(('Input number value %q for %q field does ' ..
+                'not fit in int32'):format(value, data_spec['name']))
+        end
+        check_integer(data_spec, value)
+        return encoder.encode_I32(data_spec, value)
+    elseif (protobuf_I64_types[data_type] and
+            input_I64_types[type(value)]) then
+        if (type(value) == 'number' and (value >= max_int64 or
+                value < -max_int64)) then
+            error(('Input number value %q for %q field does ' ..
+                'not fit in int64'):format(value, data_spec['name']))
+        end
+        check_integer(data_spec, value)
+        return encoder.encode_I64(data_spec, value)
+    else
+        error(('Field %q of %q type gets %q type value. ' ..
+            'Unsupported or colliding types'):format(data_spec['name'],
+            data_type, type(value)))
+    end
+end
+
+-- Encodes the entered data in accordance with the
+-- selected protocol into binary format.
+--
+-- Accepts a protocol created by protobuf.protocol function,
+-- a name of a message selected for encoding and
+-- the data that needs to be encoded in the following format.
+--
+-- data = {
+--     <field_name> = <value>,
+--     <...>
+-- }
+local function encode(protocol, name, data)
```

Comment:
As I see from the code, it doesn't accept a scalar type value for encoding. For example,

```lua
protobuf.protocol({}):encode('bool', true)
```

I think that it is useful for debugging and problem invetigating purposes. For example, what I can do with the `msgpack` module:

```lua
tarantool> msgpack.encode(true):hex()
---
- c3
...
```

That's convenient.

Next, I would organize it is a way similar to how scalars are defined in config's schema module. Please, check out the code [here](https://github.com/tarantool/tarantool/blob/0beee2a3e2cf7b53f42c3de7a04baf22bd3cb9e7/src/box/lua/config/utils/schema.lua#L107-L280).

I guess that it may look like so in the `protobuf` module:

```lua
local scalars = {}

scalars['double'] = {
    validate = function(data)
        if type(data) == 'number' then
            return
        end
        error(('Protocol Buffers type %q expects Lua type %q, got %q'):format(
            'double', 'double', type(data)))
    end,
    encode = wireformat.encode_i64,
}

<...>

local function is_scalar(_self, typename)
    return scalars[typename] ~= nil
end

local function is_enum(self, typename)
    <...>
end

local function is_message(self, typename)
    <...>
end

function methods.encode(self, typename, data)
    if is_scalar(self, typename) then
        <...>
    elseif is_enum(self, typename) then
        <...>
    elseif is_message(self, typename) then
        <...>
    else
        error(('Unknown typename %q'):format(typename))
    end
end
```

Reply:
@Col-Waltz highlighted in a f2f discussion that it is not possible to encode just scalar (such as `bool`) without its ID. A message should be on the top level. So, I discard this part of the proposal.

However, the part that describes how to better arrange the code still relevant with one fixup.

```lua
local function encode_impl(self, typename, id, data)
    if is_scalar(self, typename) then
        <...>
    elseif is_enum(self, typename) then
        <...>
    elseif is_message(self, typename) then
        <..calls encode_impl..>
    else
        error(('Unknown typename %q'):format(typename))
    end
end

function methods.encode(self, typename, data)
    <..call encode_impl..>
end
```

It seems, we need a separate top level function for a top level message, then go to a recursion in a helper function. I usually use the `_impl` suffix for such helper functions. Sometimes it is called like `encode_r` to highlight that it is recursive. (Personally I like `_impl` more.)

Reply:
It is possible that if we'll make the field encoding as a recursive function, things will be more clear.

```lua
local function encode_field(self, field_def, data)
    <...>
end
```

I can't say for sure. I propose to check different variants to organize this code and choose one that fits the best.

Reply:
Implemented these changes need to be checked

Reply:
I think that now it is solved or at least generally solved. I'll add separate comments for particular pieces of the code if there were ideas regarding them. Thanks!